DOC-13648 document magma flusher threads 7.6 (#4079)
couchbase/docs-server: modules/rest-api/pages/rest-reader-writer-thread-config.adoc

= Setting Storage Thread Allocations
:description: Couchbase Server has several settings that let you change how it allocates and uses threads for storage across the entire cluster.
:page-topic-type: reference

[abstract]
{description}

== HTTP method and URI

.Get Global Thread Settings
[source,url]
----
GET /pools/default/settings/memcached/global
----

.Set Global Thread Allocations
[source]
----
POST /pools/default/settings/memcached/global
----

[#description]
== Description

Couchbase Server lets you set the number of threads per node for reading, writing, NonIO, and AuxIO thread pools.
You can also change the number of threads that Magma buckets use for writing data to disk.
Increasing thread allocation can improve performance on systems with a large number of CPU cores.
For example, more writer threads can optimize durable write performance.

> See xref:learn:data/durability.adoc[Durability] for details.

Reviewer: ```suggestion
For more information, see xref:learn:data/durability.adoc[Durability].
```

=== Required Privileges

Your account must have at least 1 of the following roles to make a GET request from this endpoint:

* xref:learn:security/roles.adoc#full-admin[Full Admin]
* xref:learn:security/roles.adoc#cluster-admin[Cluster Admin]
* xref:learn:security/roles.adoc#external-user-security-admin[External User Security Admin]
* xref:learn:security/roles.adoc#local-user-security-admin[Local User Security Admin]

=== Responses

200 OK::
The request was successful.
Returns a JSON object containing the current global thread settings that have been set explicitly.

400 Bad Request::
The request could not be understood.
Also returns a JSON object containing an error message.

=== Example

The following example shows how to get the current global thread settings for Couchbase Server:

[source,bash]
----
curl -X GET -u Administrator:password \
http://localhost:8091/pools/default/settings/memcached/global | jq
----

If successful, the call returns a JSON object containing the any overrides to the default global thread settings:

[source,json]
----
{
  "magma_flusher_thread_percentage": 30,
  "num_storage_threads": 25
}
----

== Setting Global Thread Allocations

> To set the global thread allocations for Couchbase Server, perform a `POST` ro the `/pools/default/settings/memcached/global` endpoint.

Reviewer: ```suggestion
To set the global thread allocations for Couchbase Server, send a `POST` to the `/pools/default/settings/memcached/global` endpoint.
```

[#curl-syntax]
=== curl Syntax

[source,bash]
----

> curl -X POST http[s]://{HOST}:{PORT}/pools/default/settings/memcached/global

Reviewer: Did you want to add the $ to HOST and PORT?

Author: Our stadard's [REST API Path Attribute Placeholders](https://docs.couchbase.com/styleguide/placeholders.html#rest-api-path-attribute-placeholders) entry doesn't have us using $, like we do for BASH shell commands. These are suppose to mimic the the [OpenAPI specifications](https://swagger.io/docs/specification/v3_0/describing-parameters/) for formatting path parameters.

However, the OpenAPI standards don't use all caps for the path parameters. I'll lowercase these, and add a PR for the standards to state we should use lowercase, and add a link to the OpenAPI spec for further reference.

Reviewer: Appreciate the explanation! Might be worth updating the style guide for this too.

  [-d num_reader_threads=<integer>]
  [-d num_writer_threads=<integer>]
  [-d num_nonio_threads=<integer>]
  [-d num_auxio_threads=<integer>]
  [-d num_storage_threads=<integer>]
  [-d magma_flusher_thread_percentage=<integer>]
  -u $USER:$PASSWORD
----

 

==== Path Parameters
:required-privileges: required-privileges-put
include::partial$user_pwd_host_port_params.adoc[]

==== Parameters
(optional) Sets the number of threads Couchbase Server uses to read data.

`num_writer_threads`:: 
(optional) Sets the number of threads Couchbase Server uses to write data.

> number threads Couchbase Server uses to write data.

Reviewer: ```suggestion

```

`num_nonio_threads`:: 
(optional) Sets the number of NonIO threads to be used by Couchbase Server.
The NonIO thread pool runs in-memory tasks.
For example, the durability timeout task uses these threads.
Valid values are:

+
* An integer from `1` to `64`: sets the number of threads.
* `default`: Couchbase Server calculates and applies an appropriate number of threads.

`num_auxio_threads`:: 
(optional) Sets the number of AuxIO threads to be used by Couchbase Server.
The AuxIO thread pool runs auxiliary I/O tasks -- for example, the access log task.
Valid values are:

+
* An integer from `1` to `64`: sets the number of threads.
* `default`: Couchbase Server determines the number of threads.

`num_storage_threads`::
(optional) Sets the number of threads the Magma storage engine can use when compacting and writing data to disk.
This thread pool is divided into 2 types of threads: flusher threads and compactor threads.

> See xref:learn:buckets-memory-and-storage/storage-settings.adoc#magma-flushing-and-compaction-threads[Magma Flushing and Compaction Threads] for more information.

Reviewer: ```suggestion
For more information, see xref:learn:buckets-memory-and-storage/storage-settings.adoc#magma-flushing-and-compaction-threads[Magma Flushing and Compaction Threads].
```

+
The default value is `20`.

`magma_flusher_thread_percentage`::
(optional) Sets the percentage of Magma storage engine threads to allocate to flushing compacted data to disk.
The remaining threads in the thread pool are allocated to compacting data.
For example, if you set `num_storage_threads` to `20` and set `magma_flusher_thread_percentage` to `50`, Couchbase Server uses `10` threads to flush data and `10` threads to compact data.

> See xref:learn:buckets-memory-and-storage/storage-settings.adoc#magma-flushing-and-compaction-threads[Magma Flushing and Compaction Threads] for more information.

Reviewer: ```suggestion
For more information, see xref:learn:buckets-memory-and-storage/storage-settings.adoc#magma-flushing-and-compaction-threads[Magma Flushing and Compaction Threads].
```

NOTE: `num_storage_threads` and `magma_flusher_thread_percentage` are advanced settings.
Contact Couchbase Support before making changes to them.
Support can help you determine the best settings for your workload and hardware.

[#required-privileges-put]
=== Required Privileges

Your account must have at least 1 of the following roles to make a POST request to this endpoint:

* xref:learn:security/roles.adoc#full-admin[Full Admin]
* xref:learn:security/roles.adoc#cluster-admin[Cluster Admin]

[#responses]
=== Responses

202 Accepted::
The request was successful.
This result also returns a JSON object containing all non-default thread settings.

400 Bad Request::
The request had errors or invalid values.
This result also returns a JSON object containing an error message.

Success returns an object whose values confirm the settings that have been made.

[#examples]
=== Examples

To set the numbers of reader, writer, NonIO, and AuxIO threads for Couchbase Server, use the `POST /pools/default/settings/memcached/global` http method and endpoint as follows:

[source,bash]
----
curl -X POST -u Administrator:password \

> http://localhost:8091/pools/default/settings/memcached/global \

Reviewer: Did you want to turn these into variables instead of `localhost` and the port number?

Author: Not for an example, which is a literal command you could type and get a response from.

-d num_reader_threads=12 \
-d num_writer_threads=8 \
-d num_nonio_threads=6 \
-d num_auxio_threads=6 | jq
----

This sets the number of `reader` threads to `12`, the number of `writer` threads to `8`, and the numbers of `NonIO` and `AuxIO` threads each to `6`.
If successful, the call returns an object that confirms the settings you specified:

[source,json]
----
{
  "num_reader_threads": 12,
  "num_writer_threads": 8,
  "num_auxio_threads": 6,
  "num_nonio_threads": 6
}
----

The following example increases the Magma storage engine's thread pool to 30 threads and allocates 25% of the threads to flushing data to disk:

[source,bash]
----
curl -X POST -u Administrator:password \
     http://localhost:8091/pools/default/settings/memcached/global \
     -d num_storage_threads=30 \
     -d magma_flusher_thread_percentage=25 | jq
----

If successful, the call returns an object confirming your new settings:

[source,json]
----
{
  "magma_flusher_thread_percentage": 25,
  "num_storage_threads": 30
}
----

[#see-also]
== See Also

* See xref:learn:data/durability.adoc[Durability], for information about durable writes.

> * See xref:learn:buckets-memory-and-storage/storage-settings.adoc#threading[Threading] for an overview of threads.

Reviewer: ```suggestion
* For more information about durable writes, see xref:learn:data/durability.adoc[Durability].
* For an overview of threads, see xref:learn:buckets-memory-and-storage/storage-settings.adoc#threading[Threading].
```

Reviewer: Though, really it's okay to just have these as bare links at the end of a section with no introductory text.
